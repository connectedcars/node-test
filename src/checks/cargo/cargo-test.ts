import { filterDuplicates } from '../../common'
import { CheckAnnotation, CheckRunCompleted } from '../checks-common'
import { createAnnotationFromManifestError } from './cargo'
import { CargoMessage, CargoTestMessage } from './cargo-types'

export interface CargoTestInput {
  sha: string
  data: CargoMessage[]
}

function getTestAnnotations(item: CargoTestMessage): CheckAnnotation[] {
  // JSON output from `cargo test` is currently
  // unstable, and the path to the test case or
  // test suite is not emitted.
  // While `cargo test` does emit "compiler-artifact"
  // messages, these cannot be used to reliably infer
  // the test case or test suite's path.
  const relPath = item.name
  return [
    {
      path: relPath,
      start_line: 0,
      end_line: 0,
      annotation_level: 'failure',
      message: item.stdout ?? item.name,
      title: item.name,
      raw_details: JSON.stringify(item, null, 4)
    }
  ]
}

function getAnnotations(data: CargoMessage[]): CheckAnnotation[] {
  const annotations: CheckAnnotation[] = []
  for (const item of data) {
    if (item.reason === 'manifest-parse-error') {
      annotations.push(createAnnotationFromManifestError(item))
    } else if (item.type === 'test' && item.event === 'failed') {
      annotations.push(...getTestAnnotations(item))
    }
  }
  return annotations
}

export function cargoTestCheck({ data, sha }: CargoTestInput): CheckRunCompleted {
  if (Array.isArray(data)) {
    let annotations = getAnnotations(data)
    if (annotations.length > 0) {
      annotations = filterDuplicates(annotations)

      const summary = `Total of ${annotations.length} ${annotations.length === 1 ? 'issue' : 'issues'}`
      return {
        name: 'cargo-test',
        head_sha: sha,
        conclusion: 'failure',
        status: 'completed',
        completed_at: new Date().toISOString(),
        output: {
          title: summary,
          summary,
          annotations
        }
      }
    }
    return {
      conclusion: 'success',
      name: 'cargo-test',
      head_sha: sha,
      status: 'completed',
      completed_at: new Date().toISOString(),
      output: {
        title: 'Tests passed',
        summary: 'Tests passed'
      }
    }
  }
  return {
    conclusion: 'skipped',
    name: 'cargo-test',
    head_sha: sha,
    status: 'completed',
    completed_at: new Date().toISOString(),
    output: {
      title: 'Unexpected test output',
      summary: ''
    }
  }
}
