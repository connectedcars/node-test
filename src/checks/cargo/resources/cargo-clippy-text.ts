import { CargoMessage } from '../cargo-types'

export const cargoClippyFailedOutput: CargoMessage[] = [
  {
    reason: 'compiler-message',
    package_id: 'can 0.1.0 (path+file:///app/can)',
    target: {
      kind: ['lib'],
      crate_types: ['lib'],
      name: 'can',
      src_path: '/app/can/src/lib.rs',
      edition: '2018',
      doctest: true
    },
    message: {
      rendered:
        'error: this comparison involving the minimum or maximum element for this type contains a case that is always true or always false\n   --> can/src/can_session.rs:391:12\n    |\n391 |         if 100 > i32::MAX {}\n    |            ^^^^^^^^^^^^^^\n    |\n    = note: `-D clippy::absurd-extreme-comparisons` implied by `-D clippy::all`\n    = help: because `i32::MAX` is the maximum value for this type, this comparison is always false\n    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#absurd_extreme_comparisons\n\n',
      children: [
        {
          children: [],
          code: null,
          level: 'note',
          message: '`-D clippy::absurd-extreme-comparisons` implied by `-D clippy::all`',
          rendered: null,
          spans: []
        },
        {
          children: [],
          code: null,
          level: 'help',
          message: 'because `i32::MAX` is the maximum value for this type, this comparison is always false',
          rendered: null,
          spans: []
        },
        {
          children: [],
          code: null,
          level: 'help',
          message:
            'for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#absurd_extreme_comparisons',
          rendered: null,
          spans: []
        }
      ],
      code: {
        code: 'clippy::absurd_extreme_comparisons',
        explanation: null
      },
      level: 'error',
      message:
        'this comparison involving the minimum or maximum element for this type contains a case that is always true or always false',
      spans: [
        {
          byte_end: 13338,
          byte_start: 13324,
          column_end: 26,
          column_start: 12,
          expansion: null,
          file_name: 'can/src/can_session.rs',
          is_primary: true,
          label: null,
          line_end: 391,
          line_start: 391,
          suggested_replacement: null,
          suggestion_applicability: null,
          text: [
            {
              highlight_end: 26,
              highlight_start: 12,
              text: '        if 100 > i32::MAX {}'
            }
          ]
        }
      ]
    }
  },
  {
    reason: 'compiler-message',
    package_id: 'can 0.1.0 (path+file:///app/can)',
    target: {
      kind: ['lib'],
      crate_types: ['lib'],
      name: 'can',
      src_path: '/app/can/src/lib.rs',
      edition: '2018',
      doctest: true
    },
    message: {
      rendered: 'error: aborting due to previous error\n\n',
      children: [],
      code: null,
      level: 'error',
      message: 'aborting due to previous error',
      spans: []
    }
  }
]
