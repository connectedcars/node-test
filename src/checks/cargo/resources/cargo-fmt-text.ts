import { CargoFmtFile } from '../cargo-types'

export const cargoFmtFailedOutput: CargoFmtFile[] = [
  {
    name: '/checks-test/src/foo/bar.rs',
    mismatches: [
      {
        original_begin_line: 1,
        original_end_line: 7,
        expected_begin_line: 1,
        expected_end_line: 2,
        original: '\n\npub fn bar(\n\n)\n{\n    println!  (  "bar"  )  ;',
        expected: 'pub fn bar() {\n    println!("bar");'
      }
    ]
  },
  {
    name: '/checks-test/src/foo/mod.rs',
    mismatches: [
      {
        original_begin_line: 1,
        original_end_line: 5,
        expected_begin_line: 1,
        expected_end_line: 1,
        original: '\n\nmod bar\n;\n',
        expected: 'mod bar;'
      }
    ]
  },
  {
    name: '/checks-test/src/main.rs',
    mismatches: [
      {
        original_begin_line: 9,
        original_end_line: 9,
        expected_begin_line: 9,
        expected_end_line: 9,
        original: '',
        expected: 'mod foo;'
      },
      {
        original_begin_line: 10,
        original_end_line: 10,
        expected_begin_line: 11,
        expected_end_line: 12,
        original: ' mod foo  ;',
        expected: 'fn main() {\n    let foo = foo();'
      },
      {
        original_begin_line: 12,
        original_end_line: 16,
        expected_begin_line: 14,
        expected_end_line: 14,
        original: '\nfn main()\n{\n      let foo = foo();\n',
        expected: ''
      },
      {
        original_begin_line: 19,
        original_end_line: 19,
        expected_begin_line: 16,
        expected_end_line: 16,
        original: '',
        expected: ''
      },
      {
        original_begin_line: 27,
        original_end_line: 28,
        expected_begin_line: 23,
        expected_end_line: 23,
        original: '\n',
        expected: ''
      },
      {
        original_begin_line: 35,
        original_end_line: 38,
        expected_begin_line: 29,
        expected_end_line: 32,
        original: '\n\n    if { let b = false; b } {}\n',
        expected: '    if {\n        let b = false;\n        b\n    } {}'
      },
      {
        original_begin_line: 56,
        original_end_line: 56,
        expected_begin_line: 50,
        expected_end_line: 50,
        original: '    assert!(false);',
        expected: '        assert!(false);'
      },
      {
        original_begin_line: 71,
        original_end_line: 72,
        expected_begin_line: 65,
        expected_end_line: 65,
        original: '\n',
        expected: ''
      },
      {
        original_begin_line: 74,
        original_end_line: 75,
        expected_begin_line: 66,
        expected_end_line: 66,
        original: '\n',
        expected: ''
      }
    ]
  }
]
