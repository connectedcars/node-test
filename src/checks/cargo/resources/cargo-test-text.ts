import { CargoMessage } from '../cargo-types'

export const cargoTestFailedOutput: CargoMessage[] = [
  {
    reason: 'compiler-message',
    package_id: 'can 0.1.0 (path+file:///app/can)',
    target: {
      kind: ['lib'],
      crate_types: ['lib'],
      name: 'can',
      src_path: '/app/can/src/lib.rs',
      edition: '2018',
      doctest: true
    },
    message: {
      rendered:
        'error: expected one of `.`, `;`, `?`, `}`, or an operator, found `assert_eq`\n   --> can/src/can_session.rs:536:9\n    |\n535 |         assert!(false) \n    |                       - expected one of `.`, `;`, `?`, `}`, or an operator\n536 |         assert_eq!(session.typ, SessionType::TP2(0x89));\n    |         ^^^^^^^^^ unexpected token\n\n',
      children: [],
      code: null,
      level: 'error',
      message: 'expected one of `.`, `;`, `?`, `}`, or an operator, found `assert_eq`',
      spans: [
        {
          byte_end: 17846,
          byte_start: 17846,
          column_end: 23,
          column_start: 23,
          expansion: null,
          file_name: 'can/src/can_session.rs',
          is_primary: false,
          label: 'expected one of `.`, `;`, `?`, `}`, or an operator',
          line_end: 535,
          line_start: 535,
          suggested_replacement: null,
          suggestion_applicability: null,
          text: [
            {
              highlight_end: 23,
              highlight_start: 23,
              text: '        assert!(false) '
            }
          ]
        },
        {
          byte_end: 17865,
          byte_start: 17856,
          column_end: 18,
          column_start: 9,
          expansion: null,
          file_name: 'can/src/can_session.rs',
          is_primary: true,
          label: 'unexpected token',
          line_end: 536,
          line_start: 536,
          suggested_replacement: null,
          suggestion_applicability: null,
          text: [
            {
              highlight_end: 18,
              highlight_start: 9,
              text: '        assert_eq!(session.typ, SessionType::TP2(0x89));'
            }
          ]
        }
      ]
    }
  },
  {
    reason: 'compiler-message',
    package_id: 'can 0.1.0 (path+file:///app/can)',
    target: {
      kind: ['lib'],
      crate_types: ['lib'],
      name: 'can',
      src_path: '/app/can/src/lib.rs',
      edition: '2018',
      doctest: true
    },
    message: {
      rendered: 'error: aborting due to previous error; 2 warnings emitted\n\n',
      children: [],
      code: null,
      level: 'error',
      message: 'aborting due to previous error; 2 warnings emitted',
      spans: []
    }
  },
  {
    reason: 'build-finished',
    success: false
  },
  {
    type: 'test',
    name: 'can_session::tests::tp2_stays_alive_after_applayer_error',
    event: 'failed',
    stdout:
      "thread 'can_session::tests::tp2_stays_alive_after_applayer_error' panicked at 'assertion failed: false', can/src/can_session.rs:535:9\nnote: run with `RUST_BACKTRACE=1` environment variable to display a backtrace\n"
  },
  {
    type: 'test',
    event: 'started',
    name: 'passive_listeners::tester_detector::tests::obd2'
  },
  {
    type: 'test',
    name: 'passive_listeners::tester_detector::tests::detects_frames_on_diagnostic_relevant_arbitration_ids_sent_by_someone_else',
    event: 'ok'
  },
  {
    type: 'test',
    event: 'started',
    name: 'passive_listeners::tester_detector::tests::obd2_but_not_functionally_addressed'
  },
  {
    type: 'test',
    name: 'passive_listeners::tester_detector::tests::doesnt_hardcode_obd2',
    event: 'ok'
  },
  {
    type: 'test',
    event: 'started',
    name: 'server::tests::it_deserialize_null_sessions'
  },
  {
    type: 'test',
    name: 'passive_listeners::tester_detector::tests::functionally_addressed_but_not_obd2',
    event: 'ok'
  },
  {
    type: 'test',
    event: 'started',
    name: 'server::tests::it_deserialize_sessions_with_numbered_level'
  },
  {
    type: 'test',
    name: 'passive_listeners::tester_detector::tests::obd2',
    event: 'ok'
  },
  {
    type: 'test',
    event: 'started',
    name: 'server::tests::it_fails_empty_sessions'
  },
  {
    type: 'test',
    name: 'passive_listeners::tester_detector::tests::obd2_but_not_functionally_addressed',
    event: 'ok'
  },
  {
    type: 'test',
    event: 'started',
    name: 'server::tests::it_fails_malformed_ecu_formats'
  },
  {
    type: 'test',
    name: 'server::tests::it_deserialize_null_sessions',
    event: 'ok'
  },
  {
    type: 'test',
    event: 'started',
    name: 'server::tests::it_fails_nonsense_sessions'
  },
  {
    type: 'test',
    name: 'server::tests::it_deserialize_sessions_with_numbered_level',
    event: 'ok'
  },
  {
    type: 'test',
    event: 'started',
    name: 'server::tests::it_works_on_wellformed_ecus'
  },
  {
    type: 'test',
    name: 'server::tests::it_fails_empty_sessions',
    event: 'ok'
  },
  {
    type: 'test',
    event: 'started',
    name: 'transport::isotp::transfer_in::test::check_sn_according_to_fiat'
  },
  {
    type: 'test',
    name: 'server::tests::it_fails_malformed_ecu_formats',
    event: 'ok'
  },
  {
    type: 'test',
    event: 'started',
    name: 'transport::isotp::transfer_in::test::check_sn_according_to_iso'
  },
  {
    type: 'test',
    name: 'server::tests::it_fails_nonsense_sessions',
    event: 'ok'
  },
  {
    type: 'test',
    event: 'started',
    name: 'transport::isotp::transfer_in::test::test_flow_control'
  },
  {
    type: 'test',
    name: 'server::tests::it_works_on_wellformed_ecus',
    event: 'ok'
  },
  {
    type: 'test',
    event: 'started',
    name: 'transport::isotp::transfer_in::test::test_frame_type_from_pci1'
  },
  {
    type: 'test',
    name: 'transport::isotp::transfer_in::test::check_sn_according_to_fiat',
    event: 'ok'
  },
  {
    type: 'test',
    event: 'started',
    name: 'transport::isotp::transfer_in::test::test_handle_consecutive_frame'
  },
  {
    type: 'test',
    name: 'transport::isotp::transfer_in::test::check_sn_according_to_iso',
    event: 'ok'
  },
  {
    type: 'test',
    event: 'started',
    name: 'transport::isotp::transfer_in::test::test_handle_first_frame'
  },
  {
    type: 'test',
    name: 'transport::isotp::transfer_in::test::test_flow_control',
    event: 'ok'
  },
  {
    type: 'test',
    event: 'started',
    name: 'transport::isotp::transfer_in::test::test_handle_first_frame_unexpected'
  },
  {
    type: 'test',
    name: 'transport::isotp::transfer_in::test::test_handle_consecutive_frame',
    event: 'ignored'
  },
  {
    type: 'test',
    event: 'started',
    name: 'transport::isotp::transfer_in::test::test_handle_single_frame'
  },
  {
    type: 'test',
    name: 'transport::isotp::transfer_in::test::test_frame_type_from_pci1',
    event: 'ok'
  },
  {
    type: 'test',
    event: 'started',
    name: 'transport::isotp::transfer_in::test::test_handle_single_frame_invalid_length'
  },
  {
    type: 'test',
    name: 'transport::isotp::transfer_in::test::test_handle_first_frame',
    event: 'ok'
  },
  {
    type: 'test',
    event: 'started',
    name: 'transport::isotp::transfer_in::test::test_handle_single_frame_unexpected'
  },
  {
    type: 'test',
    name: 'transport::isotp::transfer_in::test::test_handle_first_frame_unexpected',
    event: 'ok'
  },
  {
    type: 'test',
    event: 'started',
    name: 'transport::isotp::transfer_in::test::test_is_frame_for_us'
  },
  {
    type: 'test',
    name: 'transport::isotp::transfer_in::test::test_handle_single_frame',
    event: 'ok'
  },
  {
    type: 'test',
    event: 'started',
    name: 'transport::isotp::transfer_in::test::test_waiting_for_first_or_single_frame_timeout'
  },
  {
    type: 'test',
    name: 'transport::isotp::transfer_in::test::test_handle_single_frame_invalid_length',
    event: 'ok'
  },
  {
    type: 'test',
    event: 'started',
    name: 'transport::isotp::transfer_out::test::test_send'
  },
  {
    type: 'test',
    name: 'transport::isotp::transfer_in::test::test_handle_single_frame_unexpected',
    event: 'ok'
  },
  {
    type: 'test',
    event: 'started',
    name: 'transport::isotp::transfer_out::test::test_send_not_ready'
  },
  {
    type: 'test',
    name: 'transport::isotp::transfer_in::test::test_is_frame_for_us',
    event: 'ok'
  },
  {
    type: 'test',
    event: 'started',
    name: 'transport::isotp::transfer_out::test::test_single_frame'
  },
  {
    type: 'test',
    name: 'transport::isotp::transfer_in::test::test_waiting_for_first_or_single_frame_timeout',
    event: 'ok'
  },
  {
    type: 'test',
    name: 'transport::isotp::transfer_out::test::test_send',
    event: 'ok'
  },
  {
    type: 'test',
    name: 'transport::isotp::transfer_out::test::test_send_not_ready',
    event: 'ok'
  },
  {
    type: 'test',
    name: 'transport::isotp::transfer_out::test::test_single_frame',
    event: 'ok'
  },
  {
    type: 'suite',
    event: 'failed',
    passed: 66,
    failed: 1,
    allowed_fail: 0,
    ignored: 1,
    measured: 0,
    filtered_out: 0
  }
]
