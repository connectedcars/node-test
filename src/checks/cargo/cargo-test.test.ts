import { cargoTestCheck } from './cargo-test'
import { cargoTestFailedOutput } from './resources/cargo-test-text'

describe('checks/cargo-tests', () => {
  it('converts no failed tests', () => {
    const results = cargoTestCheck({ data: [], sha: '11963e3cb7ecbb9247f638cc0fb047173a62cf7a' })

    expect(results).toEqual({
      name: 'cargo test',
      status: 'completed',
      completed_at: expect.any(String),
      head_sha: '11963e3cb7ecbb9247f638cc0fb047173a62cf7a',
      conclusion: 'success',
      output: {
        title: 'Tests passed',
        summary: 'Tests passed'
      }
    })
  })

  it('converts invalid input', () => {
    const results = cargoTestCheck({ data: {} as any, sha: '11963e3cb7ecbb9247f638cc0fb047173a62cf7a' })
    expect(results).toEqual({
      name: 'cargo test',
      status: 'completed',
      completed_at: expect.any(String),
      head_sha: '11963e3cb7ecbb9247f638cc0fb047173a62cf7a',
      conclusion: 'skipped',
      output: {
        title: 'Unexpected test output',
        summary: ''
      }
    })
  })

  it('converts failed compilation and tests', () => {
    const results = cargoTestCheck({ data: cargoTestFailedOutput, sha: '11963e3cb7ecbb9247f638cc0fb047173a62cf7a' })

    expect(results).toEqual({
      name: 'cargo test',
      status: 'completed',
      completed_at: expect.any(String),
      head_sha: '11963e3cb7ecbb9247f638cc0fb047173a62cf7a',
      conclusion: 'failure',
      output: {
        title: 'Total of 3 issues',
        summary: 'Total of 3 issues',
        annotations: [
          {
            path: 'can/src/lib.rs',
            start_line: 535,
            end_line: 535,
            annotation_level: 'failure',
            message: 'expected one of `.`, `;`, `?`, `}`, or an operator, found `assert_eq`',
            raw_details:
              '{\n' +
              '    "rendered": "error: expected one of `.`, `;`, `?`, `}`, or an operator, found `assert_eq`\\n   --> can/src/can_session.rs:536:9\\n    |\\n535 |         assert!(false) \\n    |                       - expected one of `.`, `;`, `?`, `}`, or an operator\\n536 |         assert_eq!(session.typ, SessionType::TP2(0x89));\\n    |         ^^^^^^^^^ unexpected token\\n\\n",\n' +
              '    "children": [],\n' +
              '    "code": null,\n' +
              '    "level": "error",\n' +
              '    "message": "expected one of `.`, `;`, `?`, `}`, or an operator, found `assert_eq`",\n' +
              '    "spans": [\n' +
              '        {\n' +
              '            "byte_end": 17846,\n' +
              '            "byte_start": 17846,\n' +
              '            "column_end": 23,\n' +
              '            "column_start": 23,\n' +
              '            "expansion": null,\n' +
              '            "file_name": "can/src/can_session.rs",\n' +
              '            "is_primary": false,\n' +
              '            "label": "expected one of `.`, `;`, `?`, `}`, or an operator",\n' +
              '            "line_end": 535,\n' +
              '            "line_start": 535,\n' +
              '            "suggested_replacement": null,\n' +
              '            "suggestion_applicability": null,\n' +
              '            "text": [\n' +
              '                {\n' +
              '                    "highlight_end": 23,\n' +
              '                    "highlight_start": 23,\n' +
              '                    "text": "        assert!(false) "\n' +
              '                }\n' +
              '            ]\n' +
              '        },\n' +
              '        {\n' +
              '            "byte_end": 17865,\n' +
              '            "byte_start": 17856,\n' +
              '            "column_end": 18,\n' +
              '            "column_start": 9,\n' +
              '            "expansion": null,\n' +
              '            "file_name": "can/src/can_session.rs",\n' +
              '            "is_primary": true,\n' +
              '            "label": "unexpected token",\n' +
              '            "line_end": 536,\n' +
              '            "line_start": 536,\n' +
              '            "suggested_replacement": null,\n' +
              '            "suggestion_applicability": null,\n' +
              '            "text": [\n' +
              '                {\n' +
              '                    "highlight_end": 18,\n' +
              '                    "highlight_start": 9,\n' +
              '                    "text": "        assert_eq!(session.typ, SessionType::TP2(0x89));"\n' +
              '                }\n' +
              '            ]\n' +
              '        }\n' +
              '    ]\n' +
              '}'
          },
          {
            path: 'can/src/lib.rs',
            start_line: 0,
            end_line: 0,
            annotation_level: 'failure',
            message: 'aborting due to previous error; 2 warnings emitted',
            raw_details:
              '{\n' +
              '    "rendered": "error: aborting due to previous error; 2 warnings emitted\\n\\n",\n' +
              '    "children": [],\n' +
              '    "code": null,\n' +
              '    "level": "error",\n' +
              '    "message": "aborting due to previous error; 2 warnings emitted",\n' +
              '    "spans": []\n' +
              '}'
          },
          {
            path: 'can/src/can_session.rs',
            start_line: 0,
            end_line: 0,
            annotation_level: 'failure',
            message: 'can_session::tests::tp2_stays_alive_after_applayer_error',
            raw_details:
              '{\n' +
              '    "type": "test",\n' +
              '    "name": "can_session::tests::tp2_stays_alive_after_applayer_error",\n' +
              '    "event": "failed",\n' +
              '    "exec_time": "0.000s",\n' +
              '    "stdout": "thread \'can_session::tests::tp2_stays_alive_after_applayer_error\' panicked at \'assertion failed: false\', can/src/can_session.rs:535:9\\nnote: run with `RUST_BACKTRACE=1` environment variable to display a backtrace\\n"\n' +
              '}'
          }
        ]
      }
    })
  })
})
