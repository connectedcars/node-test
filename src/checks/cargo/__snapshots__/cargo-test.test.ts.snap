// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`checks/cargo-test converts failed compilation and tests 1`] = `
Object {
  "completed_at": Any<String>,
  "conclusion": "failure",
  "head_sha": "11963e3cb7ecbb9247f638cc0fb047173a62cf7a",
  "name": "cargo-test",
  "output": Object {
    "annotations": Array [
      Object {
        "annotation_level": "failure",
        "end_line": 536,
        "message": "error: expected one of \`.\`, \`;\`, \`?\`, \`}\`, or an operator, found \`assert_eq\`
   --> can/src/can_session.rs:536:9
    |
535 |         assert!(false) 
    |                       - expected one of \`.\`, \`;\`, \`?\`, \`}\`, or an operator
536 |         assert_eq!(session.typ, SessionType::TP2(0x89));
    |         ^^^^^^^^^ unexpected token

",
        "path": "can/src/can_session.rs",
        "raw_details": "{
    \\"rendered\\": \\"error: expected one of \`.\`, \`;\`, \`?\`, \`}\`, or an operator, found \`assert_eq\`\\\\n   --> can/src/can_session.rs:536:9\\\\n    |\\\\n535 |         assert!(false) \\\\n    |                       - expected one of \`.\`, \`;\`, \`?\`, \`}\`, or an operator\\\\n536 |         assert_eq!(session.typ, SessionType::TP2(0x89));\\\\n    |         ^^^^^^^^^ unexpected token\\\\n\\\\n\\",
    \\"children\\": [],
    \\"code\\": null,
    \\"level\\": \\"error\\",
    \\"message\\": \\"expected one of \`.\`, \`;\`, \`?\`, \`}\`, or an operator, found \`assert_eq\`\\",
    \\"spans\\": [
        {
            \\"byte_end\\": 17846,
            \\"byte_start\\": 17846,
            \\"column_end\\": 23,
            \\"column_start\\": 23,
            \\"expansion\\": null,
            \\"file_name\\": \\"can/src/can_session.rs\\",
            \\"is_primary\\": false,
            \\"label\\": \\"expected one of \`.\`, \`;\`, \`?\`, \`}\`, or an operator\\",
            \\"line_end\\": 535,
            \\"line_start\\": 535,
            \\"suggested_replacement\\": null,
            \\"suggestion_applicability\\": null,
            \\"text\\": [
                {
                    \\"highlight_end\\": 23,
                    \\"highlight_start\\": 23,
                    \\"text\\": \\"        assert!(false) \\"
                }
            ]
        },
        {
            \\"byte_end\\": 17865,
            \\"byte_start\\": 17856,
            \\"column_end\\": 18,
            \\"column_start\\": 9,
            \\"expansion\\": null,
            \\"file_name\\": \\"can/src/can_session.rs\\",
            \\"is_primary\\": true,
            \\"label\\": \\"unexpected token\\",
            \\"line_end\\": 536,
            \\"line_start\\": 536,
            \\"suggested_replacement\\": null,
            \\"suggestion_applicability\\": null,
            \\"text\\": [
                {
                    \\"highlight_end\\": 18,
                    \\"highlight_start\\": 9,
                    \\"text\\": \\"        assert_eq!(session.typ, SessionType::TP2(0x89));\\"
                }
            ]
        }
    ]
}",
        "start_line": 536,
        "title": "expected one of \`.\`, \`;\`, \`?\`, \`}\`, or an operator, found \`assert_eq\`",
      },
      Object {
        "annotation_level": "failure",
        "end_line": 0,
        "message": "can_session::tests::tp2_stays_alive_after_applayer_error",
        "path": "can/src/can_session.rs",
        "raw_details": "{
    \\"type\\": \\"test\\",
    \\"name\\": \\"can_session::tests::tp2_stays_alive_after_applayer_error\\",
    \\"event\\": \\"failed\\",
    \\"exec_time\\": \\"0.000s\\",
    \\"stdout\\": \\"thread 'can_session::tests::tp2_stays_alive_after_applayer_error' panicked at 'assertion failed: false', can/src/can_session.rs:535:9\\\\nnote: run with \`RUST_BACKTRACE=1\` environment variable to display a backtrace\\\\n\\"
}",
        "start_line": 0,
      },
    ],
    "summary": "Total of 2 issues",
    "title": "Total of 2 issues",
  },
  "status": "completed",
}
`;

exports[`checks/cargo-test converts invalid input 1`] = `
Object {
  "completed_at": Any<String>,
  "conclusion": "skipped",
  "head_sha": "11963e3cb7ecbb9247f638cc0fb047173a62cf7a",
  "name": "cargo-test",
  "output": Object {
    "summary": "",
    "title": "Unexpected test output",
  },
  "status": "completed",
}
`;

exports[`checks/cargo-test converts no failed cargo-test 1`] = `
Object {
  "completed_at": Any<String>,
  "conclusion": "success",
  "head_sha": "11963e3cb7ecbb9247f638cc0fb047173a62cf7a",
  "name": "cargo-test",
  "output": Object {
    "summary": "Tests passed",
    "title": "Tests passed",
  },
  "status": "completed",
}
`;
